### PR TITLE
p2p: send messages with existing stream

### LINKED ISSUE #437: Use existing stream to send messages
## Expected Behavior

We can use an existing stream to send messages if an existing stream available.

## Current Behavior

Nodes are not receiving send messages from other nodes on an existing stream. If allow nodes to use an existing stream, the status message exchange fails.

## Possible Solution

Maybe check whether a stream has been written to before writing to it. We might need to store some metadata associated with the stream including a timestamp and the host id.

Maybe review libp2p documentation to determine if there is built-in solution.

## Steps to Reproduce

Uncomment the `getExistingStream` method and its implementation in the `send` method.

```
go test ./p2p -v
```

### PR DESCRIPTION
**Pull request by ryanchristo**
## Changes

Use an existing stream to send messages if an existing stream is available.

See specifcation 4.4.2

> For the purposes of communicating Polkadot messages, the dailer of the connection opens a unique substream. Optionally, the node can keep a unique substream alive for this purpose.

## Tests:
```
go test ./p2p --run TestExistingStream -v
```

### Issues:
closes #437

### CODE REVIEW

File: p2p/utils.go [hunk 1]
```diff
@@ -130,3 +134,88 @@ func saveKey(priv crypto.PrivKey, fp string) (err error) {
 	}
 	return f.Close()
 }
+
+// TODO: implement LEB128 variable-length encoding
```
Comment (ryanchristo) [inline p2p/utils.go]: addressed in #484

File: p2p/host.go [hunk 2]
```diff
@@ -160,32 +159,40 @@ func (h *host) ping(peer peer.ID) error {
...
-	_, err = s.Write(common.Uint16ToBytes(uint16(len(encMsg)))[0:1])
+	encMsg, err := msg.Encode()
 	if err != nil {
 		return err
 	}
 
+	// _, err = s.Write(common.Uint16ToBytes(uint16(len(encMsg)))[0:1])
+	// if err != nil {
+	// 	return err
+	// }
```
Comment (ansermino) [inline p2p/host.go]: Can this be removed?

File: p2p/service.go [hunk 3]
```diff
@@ -175,18 +168,49 @@ func (s *Service) handleConn(conn network.Conn) {
...
-// handleStream parses the message written to the data stream and calls the
-// associated message handler (non-status or status) based on message type
+// handleStream starts reading from the inbound message stream (substream with
+// a matching protocol id that was opened by the connected peer) and continues
+// reading until the inbound message stream is either closed or reset.
 func (s *Service) handleStream(stream network.Stream) {
+	ctx := context.Background()
+
+	// read from inbound stream
+	go s.readStream(ctx, stream)
```
Comment (ansermino) [inline p2p/service.go]: are you sure these needs its own goroutine? My expectation would be that this is called from a goroutine (as there are many simultaneous connections potentially happening). Removing it doesn't break tests, lmk what you think

File: p2p/host.go (hunk 2 above)
Comment (ryanchristo) [reply p2p/host.go]: I was going to remove it when taking a look at #484 but I can remove now.

File: p2p/service.go (hunk 3 above)
Comment (ryanchristo) [reply p2p/service.go]: I am not sure. I think we can remove... Let me take another look.

File: p2p/host.go (hunk 2 above)
Comment (ryanchristo) [reply p2p/host.go]: done

File: p2p/service.go (hunk 3 above)
Comment (ryanchristo) [reply p2p/service.go]: Yea, seems to be fine without it. Removed.

**Pull request merged**